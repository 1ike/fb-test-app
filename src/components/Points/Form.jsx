import React from "react";
import { Field, reduxForm } from "redux-form";

import _connect from "../../redux/connect";

const uid = (() => {
  let id = 5;
  return () => id++;
})();

const mapStateToProps = ({ center }) => ({ center });

export class Form extends React.Component {
  inputName = "inputPoint";

  createPoint = value => {
    const name = value[this.inputName];
    if (!name) {
      this.focusInput();
      return;
    }

    this.props.createPoint({
      id: uid(),
      name,
      placemark: {
        geometry: {
          coordinates: this.props.center
        },
        properties: {
          hintContent: name,
          balloonContent: name
        },
        options: {
          preset: "islands#blueCircleDotIconWithCaption",
          draggable: true
        }
      }
    });
    this.props.reset();
  };

  focusInput = () => {
    this.input.getRenderedComponent().focus();
  };

  render() {
    const title = "Добавить точку";
    return (
      <form
        id="newPointForm"
        className="new-point-form"
        onSubmit={this.props.handleSubmit(this.createPoint)}
      >
        <Field
          component="input"
          name={this.inputName}
          id={this.inputName}
          placeholder=" Введите новую точку маршрута"
          className="new-point-form__input"
          ref={input => {
            this.input = input;
          }}
          withRef
          autoFocus={true}
        />
        <button
          type="submit"
          className="new-point-form__submit-button"
          aria-label={title}
          title={title}
        >
          +
        </button>
      </form>
    );
  }
}

const reduxFormWrapper = reduxForm({ form: "createPoint" })(Form);

export default _connect(mapStateToProps)(reduxFormWrapper);
