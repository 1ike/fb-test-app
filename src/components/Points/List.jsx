import React from "react";
import { DragDropContext, Droppable, Draggable } from "react-beautiful-dnd";

import _connect from "../../redux/connect";

const CloseButton = props => {
  const title = "Удалить точку";
  return (
    <button
      type="button"
      className="sortable-list__delete-button"
      aria-label={title}
      title={title}
      onClick={props.onClick}
    >
      <span aria-hidden="true">&times;</span>
    </button>
  );
};

// a little function to help us with reordering the result
const reorder = (list, startIndex, endIndex) => {
  const result = Array.from(list);
  const [removed] = result.splice(startIndex, 1);
  result.splice(endIndex, 0, removed);

  return result;
};

const getItemStyle = (isDragging, draggableStyle) => ({
  // change background colour if dragging
  background: isDragging ? "lightgreen" : "grey",

  // styles we need to apply on draggables
  ...draggableStyle
});

const getListStyle = isDraggingOver => ({
  background: isDraggingOver ? "lightblue" : "lightgrey"
});

const mapStateToProps = state => {
  const props = {
    points: state.points
  };
  return props;
};

class List extends React.Component {
  onDragEnd = () => result => {
    // dropped outside the list
    if (!result.destination) {
      return;
    }

    const reoderedPoints = reorder(
      this.props.points,
      result.source.index,
      result.destination.index
    );

    this.props.reoderPoints(reoderedPoints);
  };

  deletePoint = id => () => {
    this.props.deletePoint(id);
  };

  render() {
    const points = this.props.points;
    return (
      <DragDropContext onDragEnd={this.onDragEnd()}>
        <Droppable droppableId="droppable">
          {(provided, snapshot) => (
            <div
              id="list"
              className="sortable-list"
              ref={provided.innerRef}
              style={getListStyle(snapshot.isDraggingOver)}
            >
              {points.map((point, index) => (
                <Draggable key={point.id} draggableId={point.id} index={index}>
                  {(provided, snapshot) => (
                    <div
                      id={`point_${point.id}`}
                      className="sortable-list__item"
                      ref={provided.innerRef}
                      {...provided.draggableProps}
                      {...provided.dragHandleProps}
                      style={getItemStyle(
                        snapshot.isDragging,
                        provided.draggableProps.style
                      )}
                    >
                      <span className="sortable-list__item-text">
                        {point.name}
                      </span>
                      <CloseButton onClick={this.deletePoint(point.id)} />
                    </div>
                  )}
                </Draggable>
              ))}
              {provided.placeholder}
            </div>
          )}
        </Droppable>
      </DragDropContext>
    );
  }
}

export default _connect(mapStateToProps)(List);
