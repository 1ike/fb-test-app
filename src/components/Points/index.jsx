import React from "react";

// import _connect from '../../connect';
import Form from "./Form";
import List from "./List";

export default class Points extends React.Component {
  render() {
    return (
      <div id="points">
        <Form />
        <List />
      </div>
    );
  }
}
