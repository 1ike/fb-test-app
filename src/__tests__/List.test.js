import React from "react";
import Enzyme, { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

import { List } from "../components/Points/List";

Enzyme.configure({ adapter: new Adapter() });

it("List snapshot", () => {
  const wrapper = shallow(<List />);
  expect(wrapper).toMatchSnapshot();
});
