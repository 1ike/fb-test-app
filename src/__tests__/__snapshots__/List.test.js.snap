// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <List />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <Connect(Droppable)
        direction="vertical"
        droppableId="droppable"
        ignoreContainerClipping={false}
        isDropDisabled={false}
        type="DEFAULT"
>
        [Function]
</Connect(Droppable)>,
      "onDragEnd": [Function],
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": [Function],
        "direction": "vertical",
        "droppableId": "droppable",
        "ignoreContainerClipping": false,
        "isDropDisabled": false,
        "type": "DEFAULT",
      },
      "ref": null,
      "rendered": [Function],
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <Connect(Droppable)
          direction="vertical"
          droppableId="droppable"
          ignoreContainerClipping={false}
          isDropDisabled={false}
          type="DEFAULT"
>
          [Function]
</Connect(Droppable)>,
        "onDragEnd": [Function],
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": [Function],
          "direction": "vertical",
          "droppableId": "droppable",
          "ignoreContainerClipping": false,
          "isDropDisabled": false,
          "type": "DEFAULT",
        },
        "ref": null,
        "rendered": [Function],
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
